### PR TITLE
Updated Token Handling

### PR DESCRIPTION
**Pull request by jennerl**
airmap::rest::Authenticator is now responsible for handling auth internally provided the user provided credentials in a similar manner as described in this repo's setup instructions.

### CODE REVIEW

File: include/airmap/authenticator.h [hunk 1]
```diff
@@ -110,22 +105,10 @@ class AIRMAP_EXPORT Authenticator : DoNotCopyOrMove {
   virtual void renew_authentication(const RenewAuthentication::Params& params,
                                     const RenewAuthentication::Callback& cb) = 0;
   
-  /// connect a slot to the signal which will be emitted whenever the auth token is updated.
-  boost::signals2::connection connect(const authentication_signal_t::slot_type &subscriber) {
-    return auth_signal_.connect(subscriber);
-  }
-
-  void notify_auth_token_updated(std::string token) {
```
Comment (jennerl) [inline include/airmap/authenticator.h:118]: the major change between the previous version of token handling and this is that the authenticator will now automatically, to the best of its ability, handle renewing / fetching tokens

File: src/airmap/net/http/jwt_provider.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <airmap/optional.h>
+
+namespace airmap {
+namespace net {
+namespace http {
+
+class JWTProvider {
+public:
+    virtual Optional<std::string> jwt_string() = 0;
```
Comment (jennerl) [inline src/airmap/net/http/jwt_provider.h:24]: generic mechanism to allow framework to expect a jwt_string that the service calls can supply in the header if present

File: src/airmap/rest/authenticator.cpp [hunk 3]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+    }
+
+    auto token = airmap::Token{t};
+    write_token_to_file_(token);
+  } else {
+    log_.errorf(component, "failed to authenticate with the Airmap services: %s", result.error());
+  }
+}
+
+void airmap::rest::Authenticator::renew_authentication_token(const Credentials& credentials, const Token& token) {
```
Comment (jennerl) [inline src/airmap/rest/authenticator.cpp:160]: ok, this isn't rest, specifically, but a higher order construct to renew a token...which makes getting the jwt string easier...same applies to `request_authentication_token`

File: src/airmap/rest/authenticator.cpp [hunk 4]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+    this->authenticate_with_password(
+        params, std::bind(&airmap::rest::Authenticator::handle_result_for_authentication_with_password_, this, ph::_1));
+  } else if (credentials.anonymous) {
+    this->authenticate_anonymously(
+        {credentials.anonymous.get().id},
+        std::bind(&airmap::rest::Authenticator::handle_result_for_anonymous_authentication_, this, ph::_1));
+  }
+}
+
+airmap::Optional<std::string> airmap::rest::Authenticator::jwt_string() {
```
Comment (jennerl) [inline src/airmap/rest/authenticator.cpp:187]: this is the specific implementation of getting a jwt string from AirMap's rest services. std::future used here to control concurrency so we can wait on those async network calls to finish before we try to return the jwt string

File: src/airmap/rest/client.cpp [hunk 5]
```diff
@@ -19,32 +19,39 @@ airmap::rest::Client::Client(const Configuration& configuration, const std::shar
...
       udp_sender_{sender},
       mqtt_broker_{broker},
+      authenticator_{std::make_shared<airmap::net::http::AuthorizedRequester>(configuration_.credentials.api_key,
+                                                                              requesters.authenticator,
+                                                                              nullptr),
+                                                                              requesters.sso},
       advisory_{std::make_shared<airmap::net::http::AuthorizedRequester>(configuration_.credentials.api_key,
-                                                                         requesters.advisory)},
+                                                                         requesters.advisory,
+                                                                         &authenticator_)},
```
Comment (jennerl) [inline src/airmap/rest/client.cpp:28]: all service interfaces (save the authenticator) now take an optional object that can provide a jwt string so the services can manage their token handling internally and hide this from users

File: src/airmap/rest/client.cpp [hunk 6]
```diff
@@ -90,14 +97,3 @@ airmap::Telemetry& airmap::rest::Client::telemetry() {
 airmap::Traffic& airmap::rest::Client::traffic() {
   return traffic_;
 }
-
-void airmap::rest::Client::handle_auth_update(std::string token) {
```
Comment (jennerl) [inline src/airmap/rest/client.cpp:94]: we no longer need to do this explicitly since, if we want to provide a jwt string in the header, we just ask for it and maybe get it

File: src/airmap/rest/authenticator.cpp [hunk 7]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+    auto token = read_token_from_file_();
+    if (token) {
+      if (token.get().type() == Token::Type::oauth) {
+        if (token.get().oauth().refresh.empty()) {
+          log_.errorf(component, "token file does not hold renewable token");
+        } else {
+          std::future<void> renew_future = std::async(std::launch::async, [config, token, this] {
+            renew_authentication_token(config.credentials, token.get());
+          });
+          renew_future.wait();
```
Comment (remster) [inline src/airmap/rest/authenticator.cpp:206]: this will block the current thread until the one spawned by `std::async(std::launch::async, ...` - what is the purpose of spawning an additional thread?

File: src/airmap/rest/authenticator.cpp (hunk 4 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp:187]: this logic purposefully follows the flow of `cmds::airmap::cmd::Login` as it is intended to be a one-stop shop for authentication across the platform-sdk

File: src/airmap/net/http/jwt_provider.h (hunk 2 above)
Comment (remster) [reply src/airmap/net/http/jwt_provider.h:24]: why does this return an option? JWTProvider is itself optional, why does it return an optional (as opposed to mandatory) token? Can you document the reasoning?

File: src/airmap/rest/authenticator.cpp [hunk 8]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+        {credentials.anonymous.get().id},
+        std::bind(&airmap::rest::Authenticator::handle_result_for_anonymous_authentication_, this, ph::_1));
+  }
+}
+
+airmap::Optional<std::string> airmap::rest::Authenticator::jwt_string() {
+  auto config_path = airmap::paths::config_file(Client::Version::production).string();
+  std::ifstream config_file{config_path};
+  if (!config_file) {
+    log_.errorf(component, "failed to open config file %s", config_path);
```
Comment (remster) [inline src/airmap/rest/authenticator.cpp:191]: this smells like the code is misconfigured and it should bark way louder (panic or at least throw)>

File: src/airmap/rest/authenticator.cpp [hunk 9]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+        if (!token.get().refreshed().original_token) {
+          log_.errorf(component, "token file does not hold renewable token");
+        } else {
+          std::future<void> renew_future = std::async(std::launch::async, [config, token, this] {
+            renew_authentication_token(config.credentials, token.get());
+          });
+          renew_future.wait();
+        }
+      } else {
+        log_.errorf(component, "token file does not hold renewable token");
```
Comment (remster) [inline src/airmap/rest/authenticator.cpp:218]: what can can it hold? what's the way out of this?

File: src/airmap/rest/authenticator.cpp [hunk 10]
```diff
@@ -111,3 +115,141 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+      }
+    } else {
+      log_.errorf(component, "failed to open token file for reading");
```
Comment (remster) [inline src/airmap/rest/authenticator.cpp]: maybe you should purge that (seemingly corrupt) token?

File: src/airmap/net/http/jwt_provider.h (hunk 2 above)
Comment (jennerl) [reply src/airmap/net/http/jwt_provider.h:24]: optional since the user can, for instance, not have provided credentials, or if the auth service is non-resonsive. I can comment in the code if appropriate

File: src/airmap/rest/authenticator.cpp (hunk 7 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp:206]: Honestly I struggled with this a bit. I defined the interface of `JWTProvider` to return a string so that the callers would have a simple interface to get the jwt string. I suppose that interface could change as you suggest.

The additional thread spawn may be unnecessary. I'm not versed with `std::future` but did want to block until the callback was complete...and I could extract the jwt string from the resulting token. If there's a better way to accomplish that without spawning another thread, that's a good improvement.

File: src/airmap/rest/authenticator.cpp (hunk 8 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp:191]: in `cmd::Login`, where there's an error, it stops the context. I think it would be appropriate to throw some sort of exception instead of asserting and killing the parent program

File: src/airmap/rest/authenticator.cpp (hunk 9 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp:218]: I suppose we could just request a new token in these cases

File: src/airmap/rest/authenticator.cpp (hunk 10 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp]: updated to remove token file

File: src/airmap/rest/authenticator.cpp (hunk 9 above)
Comment (jennerl) [reply src/airmap/rest/authenticator.cpp:222]: for now, I've just removed the token file. Something more able to resolve this issue would be helpful in the future...like requesting a new token

**Comment by jennerl** [comment]
This code change should be able to be tested via the daemon's command set since the updated token handling is used by the underlying services that power the command set.

So just following the normal (as documented in this repo) workflow of using the command set to log in then execute tests should be a valid test.

File: src/airmap/rest/authenticator.cpp (hunk 7 above)
Comment (tim-du) [reply src/airmap/rest/authenticator.cpp]: I removed the `std::future` stuff in authenticator.cpp. I didn't really replace it with anything - I think the caller of `jwt_string()` could make their own `std::future<std::string>` if they wanted. I don't think the old implementation had any built in async calls either.

File: src/airmap/net/http/jwt_provider.h (hunk 2 above)
Comment (tim-du) [reply src/airmap/net/http/jwt_provider.h:25]: added comment

File: src/airmap/rest/authenticator.cpp (hunk 9 above)
Comment (tim-du) [reply src/airmap/rest/authenticator.cpp]: Now requesting new token.

File: src/airmap/rest/authenticator.cpp (hunk 8 above)
Comment (tim-du) [reply src/airmap/rest/authenticator.cpp]: This throws now.

File: src/airmap/cmds/airmap/cmd/login.h [hunk 11]
```diff
@@ -31,6 +31,8 @@ namespace cmds {
 namespace airmap {
 namespace cmd {
 
+// TODO: find a way to consolidate login and authenticator code
+
 class Login : public util::cli::CommandWithFlagsAndAction {
```
Comment (tim-du) [inline src/airmap/cmds/airmap/cmd/login.h:36]: I didn't notice this before, but the new authenticator functions are really similar to the login functions. It would be nice if these were not duplicated. Although, the changes I made to authenticator mean that they are somewhat different now.

File: src/airmap/rest/authenticator.cpp [hunk 12]
```diff
@@ -111,3 +115,144 @@ void airmap::rest::Authenticator::renew_authentication(const RenewAuthentication
...
+    request_authentication_token(config.credentials, func);
+    return;
+  }
+
+  auto token_type = token.get().type();
+
+  if ((token_type != Token::Type::oauth && token_type != Token::Type::refreshed) ||
+      (token_type == Token::Type::oauth && token.get().oauth().refresh.empty()) ||
+      (token_type == Token::Type::refreshed && !token.get().refreshed().original_token)) { // TODO: what about anonymous token?
+    log_.errorf(component, "token file does not hold renewable token");
```
Comment (tim-du) [inline src/airmap/rest/authenticator.cpp:222]: There is another token type defined, which is `anonymous`. This if-clause would consider it non-renewable, and remove the token file. I'm not sure if this behavior is correct.

File: src/airmap/rest/authenticator.h [hunk 13]
```diff
@@ -35,10 +38,33 @@ class Authenticator : public airmap::Authenticator {
...
                             const RenewAuthentication::Callback& cb) override;
+  
+  void renew_authentication_token(const Credentials& credentials,
+                                  const Token& token,
+                                  std::function<void(Optional<std::string>)> cb = {});
+
+  void request_authentication_token(const Credentials& credentials,
+                                    std::function<void(Optional<std::string>)> cb = {});
+
+  void perform_with_auth(std::function<void(Optional<std::string>)>) override;
```
Comment (tim-du) [inline src/airmap/rest/authenticator.h:49]: Instead of just having a function that gets the JWT string, this function will make the request to renew authentication, and then perform some function which requires authentication once it has the JWT by calling the `std::function` which is passed to it, like a callback. I made this change so the authentication request could be asynchronous like all other requests.

File: src/airmap/cmds/airmap/cmd/login.h (hunk 11 above)
Comment (remster) [reply src/airmap/cmds/airmap/cmd/login.h:36]: Can you expand on this a bit? What's the difference between login and authenticator? Can you offer links?

File: src/airmap/rest/authenticator.cpp (hunk 12 above)
Comment (remster) [reply src/airmap/rest/authenticator.cpp:222]: I never understood what an anonymous token was.

**Comment by tim-du** [comment]
@remster here's a more detailed description:

What's been done:

- Files under cmds/airmap/cmd have just had token-related code removed.
  - They all used to read the token file to get the token, and eventually pass it to the AuthorizedRequester , which would add it to the headers for any request.
  - Instead, the AuthorizedRequester now uses the Authenticator to make authorized requests via the perform_with_auth function, which takes a callback as an argument, gets a JWT token if it doesn't already have one, and calls the callback with the token as an argument.
- It looks like the token is now in a file in a cache directory instead of the config directory (see paths.cpp changes)


Design effects:

- User no longer needs to set the auth token via the API. As long as they've successfully logged in, Authenticator should add the token to requests and renew it if necessary.
- Less duplicated code in cmds and rest


Why breaking:

- Basically everything under airmap/rest had a public member function called set_auth_token  which would just would just call the AuthorizedRequester's set_auth_token public member function. That was so AuthorizedRequester could add the token to any requests as described above. None of the set_auth_token functions exist anymore.

File: src/airmap/cmds/airmap/cmd/login.h (hunk 11 above)
Comment (tim-du) [reply src/airmap/cmds/airmap/cmd/login.h:36]: From my knowledge, login the implementation of a command (like you would use in the terminal) and authenticator is part of the API.

I say they're similar since they have a lot of the same structure now. I think originally a lot of the login code was copied to authenticator.

[See these `handle_result...` functions in login](https://github.com/airmap/platform-sdk/blob/23559cce4df5ef7cd5fd128117e202178e88635c/src/airmap/cmds/airmap/cmd/login.cpp#L190-L263)

vs [these in authenticator](https://github.com/airmap/platform-sdk/blob/23559cce4df5ef7cd5fd128117e202178e88635c/src/airmap/rest/authenticator.cpp#L119-L168)

They are somewhat different, although a part of that is that the writing to file part of the login functions is in a helper function in the authenticator code. However, the login functions use the `context` and I don't know if that makes sense at all in the authenticator versions. They just seem structurally similar and I feel like if in the future one needs to be changed, it will mean that the other should be changed as well, but could be forgotten. I would rather have the implementations combined somehow.

Comment (tim-du) [reply src/airmap/cmds/airmap/cmd/login.h:36]: Looking at it a bit more, I think I could make it so the login functions call the authenticator functions without a ton of trouble. Maybe worth doing, what do you think?

**Review by remster — APPROVED** [review]
